File: python-ai/app/services/document_types.py
import json
from pathlib import Path

CONFIG_PATH = Path(__file__).resolve().parent.parent.parent.parent / "wasm" / "document-types.json"

_doc_types_cache: dict | None = None


def get_document_types():
    global _doc_types_cache
    if _doc_types_cache is None:
        with open(CONFIG_PATH) as f:
            raw = json.load(f)

        result = {}
        for name, cfg in raw.items():
            result[name] = {
                "labels": cfg["detection"]["keywords"],
                "id_patterns": cfg["validation"]["idPatterns"],
            }
        _doc_types_cache = result

    return _doc_types_cache

File: wasm/document_types.go
package main

import (
	_ "embed"
	"encoding/json"
	"regexp"
	"sync"
)

//go:embed document-types.json
var documentTypesJSON []byte

type rawDocumentType struct {
	Label      string           `json:"label"`
	Detection  rawDetection     `json:"detection"`
	Validation rawValidation    `json:"validation"`
}

type rawDetection struct {
	Keywords []string `json:"keywords"`
	Patterns []string `json:"patterns"`
}

type rawValidation struct {
	IDPatterns []string `json:"idPatterns"`
}

var loadDocProfilesOnce sync.Once

func loadDocumentProfiles() (map[string]DocumentProfile, error) {
	var loadErr error
	loadDocProfilesOnce.Do(func() {
		var raw map[string]rawDocumentType
		if err := json.Unmarshal(documentTypesJSON, &raw); err != nil {
			loadErr = err
			return
		}

		result := make(map[string]DocumentProfile, len(raw))
		for name, r := range raw {
			profile := DocumentProfile{
				Labels: r.Detection.Keywords,
			}

			for _, p := range r.Validation.IDPatterns {
				re, err := regexp.Compile(p)
				if err != nil {
					loadErr = err
					return
				}
				profile.IDPatterns = append(profile.IDPatterns, re)
			}

			result[name] = profile
		}

		documentProfiles = result
	})

	return documentProfiles, loadErr
}

func init() {
	_, err := loadDocumentProfiles()
	if err != nil {
		panic(err)
	}
}
